pkg/security/security_profile/activity_tree/activity_tree.go

```diff
@@ -0,0 +1,378 @@
+func (at *ActivityTree) Contains(event *model.Event) (bool, error) {
+	newEntry, err := at.insert(event, true)
+	return !newEntry, err
+}
+
+// insert inserts the event in the activity tree, returns true if the event generated a new entry in the tree
+func (at *ActivityTree) insert(event *model.Event, shadowInsertion bool) (newEntry bool, outputErr error) {
+	// metrics
+	defer func() {
+		if newEntry && !shadowInsertion {
+			// this doesn't count the exec events which are counted separately
+			at.Stats.addedCount[event.GetEventType()][Runtime].Inc()
+		}
+	}()
+
+	// check if this event type is traced
+	if valid, err := at.isEventValid(event, shadowInsertion); !valid || err != nil {
+		return false, fmt.Errorf("invalid event: %s", err)
+	}
+
+	// find the node where the event should be inserted
+	entry, _ := event.FieldHandlers.ResolveProcessCacheEntry(event)
+	if entry == nil {
+		return false, fmt.Errorf("process cache entry not found")
+	}
+	if !entry.HasCompleteLineage() && !shadowInsertion { // check that the process context lineage is complete, otherwise drop it
+		at.Stats.droppedCount[event.GetEventType()][brokenLineageReason].Inc()
+		return false, fmt.Errorf("incomplete lineage")
+	}
+
+	node, newProcessNode, err := at.CreateProcessNode(entry, Runtime, shadowInsertion)
+	if newProcessNode && shadowInsertion {
+		return true, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	if node == nil {
+		// a process node couldn't be found or created for this event, ignore it
+		return false, err
+	}
+
+	// resolve fields
+	event.ResolveFieldsForAD()
+
+	// the count of processed events is the count of events that matched the activity dump selector = the events for
+	// which we successfully found a process activity node
+	at.Stats.processedCount[event.GetEventType()].Inc()
+
+	// insert the event based on its type
+	switch event.GetEventType() {
+	case model.ExecEventType:
+		// tag the matched rules if any
+		node.MatchedRules = model.AppendMatchedRule(node.MatchedRules, event.Rules)
+		return newProcessNode, nil
+	case model.FileOpenEventType:
+		return node.InsertFileEvent(&event.Open.File, event, Runtime, at.Stats, at.shouldMergePaths, shadowInsertion)
+	case model.DNSEventType:
+		return node.InsertDNSEvent(event, at.Stats, at.DNSNames, shadowInsertion)
+	case model.BindEventType:
+		return node.InsertBindEvent(event, at.Stats, shadowInsertion)
+	case model.SyscallsEventType:
+		return node.InsertSyscalls(event, at.SyscallsMask)
+	}
+
+	return false, nil
+}
+
+// CreateProcessNode finds or a create a new process activity node in the activity dump if the entry
+// matches the activity dump selector.
+func (at *ActivityTree) CreateProcessNode(entry *model.ProcessCacheEntry, generationType NodeGenerationType, shadowInsertion bool) (node *ProcessNode, newProcessNode bool, err error) {
+	if entry == nil {
+		return nil, false, nil
+	}
+
+	// drop processes with abnormal paths
+	if entry.GetPathResolutionError() != "" {
```

Comment:
same here see the line : https://github.com/DataDog/datadog-agent/pull/16744/files#diff-c7e94094f2ca1d0e250b89fa8f9cd016c90791b1c9ca72b7b376c27bc6ea4492R161